python-checks/src/main/java/org/sonar/python/checks/SklearnCachedPipelineDontAccessTransformersCheck.java

```diff
@@ -0,0 +1,209 @@
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.quickfix.PythonQuickFix;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.AssignmentStatement;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.ListLiteral;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.QualifiedExpression;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.Tuple;
+import org.sonar.plugins.python.api.tree.UnpackingExpression;
+import org.sonar.python.quickfix.TextEditUtils;
+import org.sonar.python.semantic.SymbolUtils;
+import org.sonar.python.tree.TreeUtils;
+import org.sonar.python.tree.TupleImpl;
+import org.sonar.python.types.InferredTypes;
+
+@Rule(key = "S6971")
+public class SklearnCachedPipelineDontAccessTransformersCheck extends PythonSubscriptionCheck {
+
+  public static final String MESSAGE = "Avoid accessing transformers in a cached pipeline.";
+  public static final String MESSAGE_SECONDARY = "The transformer is accessed here";
+  public static final String MESSAGE_SECONDARY_CREATION = "The Pipeline is created here";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, SklearnCachedPipelineDontAccessTransformersCheck::checkCallExpr);
+  }
+
+  private static void checkCallExpr(SubscriptionContext subscriptionContext) {
+    CallExpression callExpression = (CallExpression) subscriptionContext.syntaxNode();
+    Optional<PipelineCreation> pipelineCreationOptional = isPipelineCreation(callExpression);
+    if (pipelineCreationOptional.isEmpty()) {
+      return;
+    }
+    PipelineCreation pipelineCreation = pipelineCreationOptional.get();
+
+    var memoryArgument = TreeUtils.argumentByKeyword("memory", callExpression.arguments());
+    if (memoryArgument == null || memoryArgument.expression().is(Tree.Kind.NONE) || memoryArgument.expression().type() == InferredTypes.anyType()) {
+      return;
+    }
+    var stepsArgument = TreeUtils.nthArgumentOrKeyword(0, "steps", callExpression.arguments());
+
+    Map<Name, String> nameToStepName = new HashMap<>();
+    Optional<Expression> stepArgumentExpression = Optional.ofNullable(stepsArgument)
+      .map(RegularArgument::expression);
+
+    var stepNames = stepArgumentExpression.map(
+      e -> pipelineCreation == PipelineCreation.PIPELINE ? extractFromPipeline(e, nameToStepName) : extractFromMakePipeline(e))
+      .orElse(Stream.empty());
+
+    var qualifiedUses = stepNames
+      .map(name -> Map.entry(name, symbolIsUsedInQualifiedExpression(name)));
+
+    qualifiedUses.forEach(entry -> {
```

Comment:
In hindsight maybe we could split this method into small methods. It would make it easier to read. We can also drop the assignment to qualifiedUses. 
```suggestion
      stepNames
      .map(name -> Map.entry(name, symbolIsUsedInQualifiedExpression(name)));
      .forEach(...)
```